src/libmeasurement_kit/neubot/dash_impl.hpp

```diff
@@ -0,0 +1,345 @@
+            cb(error);
+        },
+        reactor, logger);
+}
+
+template <MK_MOCK_AS(http::request_connect, http_request_connect),
+          MK_MOCK_AS(http::request_sendrecv, http_request_sendrecv_negotiate),
+          MK_MOCK_AS(http::request_sendrecv, http_request_sendrecv_collect)>
+void negotiate_with_(std::string url, Var<report::Entry> entry,
+                     Settings settings, Var<Reactor> reactor,
+                     Var<Logger> logger, Callback<Error> cb) {
+    logger->info("Negotiating with: %s", url.c_str());
+    settings["http/url"] = url;
+    http_request_connect(
+        settings,
+        [=](Error error, Var<net::Transport> txp) {
+            if (error) {
+                // Note: in this case we don't need to close the transport
+                logger->warn("neubot: cannot connect to negotiate server: %s",
+                             error.explain().c_str());
+                cb(error);
+                return;
+            }
+            // Note: from now on, we own the `txp` transport
+            assert(!!txp);
+            negotiate_loop_<http_request_sendrecv_negotiate>(
+                entry, txp, settings, reactor, logger,
+                [=](Error error, std::string auth_token) {
+                    if (error) {
+                        logger->warn("neubot: negotiate failed: %s",
+                                     error.explain().c_str());
+                        txp->close([=]() { cb(error); });
+                        return;
+                    }
+                    // TODO: generalize code, allow to run more tests
+                    run_impl(url, auth_token, entry, settings, reactor, logger,
+                             [=](Error error) {
+                                 if (error) {
+                                     logger->warn("neubot: test failed: %s",
+                                                  error.explain().c_str());
+                                     /* FALLTHROUGH */
+                                 }
+                                 logger->info("Collecting results");
+                                 collect_<http_request_sendrecv_collect>(
+                                     txp, entry, auth_token, settings, reactor,
+                                     logger, [=](Error error) {
+                                         // Dispose of the `txp`
+                                         txp->close([=]() { cb(error); });
+                                     });
+                             });
+                });
+        },
+        reactor, logger);
+}
+
+template <MK_MOCK_AS(mlabns::query, mlabns_query)>
+void negotiate_impl(Var<report::Entry> entry, Settings settings,
+                    Var<Reactor> reactor, Var<Logger> logger,
+                    Callback<Error> cb) {
+    if (settings.find("url") != settings.end()) {
+        negotiate_with_(settings["url"], entry, settings, reactor, logger, cb);
+        return;
+    }
```

Comment:
Note: I think that, on Neubot, we usually specify address and port rather than a URL, so perhaps here it would be much better to have in input address and port rather than URL as well. (Also, the command line should be modified such that it is possible to pass address and port.)